### PR TITLE
Allow setting CDN, API URL and HTTP proxy settings in Hex config

### PR DESCRIPTION
**Pull request by JamesS237**
This pull request allows users to set the following configuration settings from the config file:

- API URL
- CDN URL
- HTTP proxy
- HTTPS proxy

>Note: Sadly, setting the home directory from the config file is impossible.

This allows users to do the following:
```shell
$ mix hex.config api_url http://hex.internal.corporation.com
$ mix hex.config cdn_url http://cdn.hex.internal.corporation.com
```

:christmas_tree: :fireworks: Merry Christmas and a Happy New Year!

### CODE REVIEW

**Comment by JamesS237** [comment]
screen shot 2015-01-01 at 4 49 55 pm
Here's the new docs rendered.

**Comment by josevalim** [comment]
Beautiful!

File: lib/hex.ex [hunk 1]
```diff
@@ -18,12 +17,24 @@ defmodule Hex do
...
+    if url  = Keyword.get(config, :api_url), do: url(url)
+    if cdn  = Keyword.get(config, :cdn_url), do: cdn(cdn)
+
+    # HTTP proxy configuration from both environment variables and Hex config
+    http_proxy  = System.get_env("HTTP_PROXY")
+                  || System.get_env("http_proxy")
+                  || Keyword.get(config, :http_proxy)
+    https_proxy = System.get_env("HTTPS_PROXY")
+                  || System.get_env("https_proxy")
+                  || Keyword.get(config, :https_proxy)
```
Comment (ericmj) [inline lib/hex.ex]: Env vars should override the config, not the other way override.

File: lib/mix/tasks/hex/config.ex [hunk 2]
```diff
@@ -7,6 +7,16 @@ defmodule Mix.Tasks.Hex.Config do
...
+
+  ## Default:
+   - `username` - Hex username
+   - `key` - Hex API key
+
+  ## Optional:
+   - `api_url` - Hex API base URL (without trailing slash)
+   - `cdn_url` - Hex CDN base URL (without trailing slash)
+   - `http_proxy` - HTTP proxy server
+   - `https_proxy` - HTTPS proxy server
```
Comment (ericmj) [inline lib/mix/tasks/hex/config.ex]: No colons in headers and I think using the word default is wrong here. What do you think about this?

```
 ## Config keys
  - `username` - Hex username
  - `key` - Hex API key
  - `api_url` - Hex API base URL (without trailing slash)
  - `cdn_url` - Hex CDN base URL (without trailing slash)
  - `http_proxy` - HTTP proxy server
  - `https_proxy` - HTTPS proxy server

Comment (ericmj) [inline lib/mix/tasks/hex/config.ex]: We should trim the trailing slash instead of documenting that it isn't allowed.

Comment (JamesS237) [inline lib/mix/tasks/hex/config.ex]: I agree on both of these points here. Fixing up and rebasing.

File: lib/hex.ex (hunk 1 above)
Comment (JamesS237) [inline lib/hex.ex]: Agreed.

File: lib/mix/tasks/hex/config.ex (hunk 2 above)
Comment (JamesS237) [inline lib/mix/tasks/hex/config.ex]: Turns out we already trim the trailing slash! Whoops

**Comment by JamesS237** [comment]
All fixed and rebased. Ready to go!

**Comment by ericmj** [comment]
Thanks James!

**Pull request merged**